fix(icon): auto assign key for each path element if createIcon prop path is array

zgid123 opened the pull request:
## 📝 Description

> Assign key for each path if we pass array of paths for createIcon

## ⛳️ Current behavior (updates)

> We will need to pass key for each path manually

```ts
import { createIcon } from '@chakra-ui/icons';

export const HeartIcon = createIcon({
  displayName: 'HeartIcon',
  path: [
    <path key={0} stroke='none' d='M0 0h24v24H0z' fill='none' />,
    <path
      key={1}
      d='M19.5 13.572l-7.5 7.428l-7.5 -7.428m0 0a5 5 0 1 1 7.5 -6.566a5 5 0 1 1 7.5 6.572'
    />,
  ],
});
```

## 🚀 New behavior

> createIcon will assign key automatically for each path, we don't need to assign when passing to the prop

```ts
import { createIcon } from '@chakra-ui/icons';

import { defaultProps } from './common';

export const HeartIcon = createIcon({
  displayName: 'HeartIcon',
  defaultProps,
  path: [
    <path stroke='none' d='M0 0h24v24H0z' fill='none' />,
    <path
      d='M19.5 13.572l-7.5 7.428l-7.5 -7.428m0 0a5 5 0 1 1 7.5 -6.566a5 5 0 1 1 7.5 6.572'
    />,
  ],
});
```

## 💣 Is this a breaking change (Yes/No): No

## 📝 Additional Information

This will fix #5127

ljosberinn commented on packages/icon/src/create-icon.tsx:
```diff
@@ -32,10 +33,14 @@ export function createIcon(options: CreateIconOptions) {
+  let path = options.path
+
+  if (Array.isArray(path)) {
+    path = path.map((p) => React.cloneElement(p, { key: uuid() }))
```
I'd refrain from adding an additional dependency for this for these reasons:
- the `path` property of the icon is highly unlikely to change. Using the array index ( `path.map((p, index) => cloneElement(p, { key: index }))` is sufficient here
- although unlikely, if `createIcon` is called multiple times, e.g. because its within the body of a rerendering component, `uuid` would re-execute and create a _new_ key for the _same_ path, which defeats the purpose of using a key in the first place.

ljosberinn commented:
Hi @zgid123 - thank you for the PR, this is a great idea! I've added a comment above :)

zgid123 commented:
@ljosberinn updated

segunadebayo commented on packages/icon/src/create-icon.tsx:
```diff
@@ -32,10 +32,14 @@ export function createIcon(options: CreateIconOptions) {
     displayName,
     defaultProps = {},
   } = options
+  let path = options.path
+
+  if (Array.isArray(path)) {
+    path = path.map((p, index) => React.cloneElement(p, { key: index }))
+  }
```
I think we can use React's internals to resolve this. Let's use `React.Children.toArray` instead.

```suggestion
  const path = React.Children.toArray(options.path)
```

ljosberinn replied:
while it is an array now, the mapping to add the key appears to be missing :)

ljosberinn replied:
interesting, TIL [code](https://github.com/facebook/react/blob/cae635054e17a6f107a39d328649137b83f25972/packages/react/src/ReactChildren.js#L112) - looks like this is good to go then!

segunadebayo reviewed (changes requested):
Thanks for working on this @zgid123.

- See my comment below.
- Please add a changeset by running `yarn changeset`

zgid123 commented:
updated

zgid123 commented:
Sorry for spaming this PR. I just updated the code again

I changed the code from
```ts
{path ?? <path fill="currentColor" d={pathDefinition} />}
```
to
```ts
{path.length ? path : <path fill="currentColor" d={pathDefinition} />}
```

Because for old logic, `path` can be `undefined`, so if path is undefined, we will render `path` tag.

But for current logic, path is always an array, if we don't pass `path`, it will be empty array instead of `undefined`, so I change to that logic, if we pass `path` to the property, it will render path, else will render `path` tag
